src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs

```diff
@@ -0,0 +1,517 @@
+                    updatedParts[i] = VisitPart(part);
+                }
+
+                return SegmentCore(updatedParts);
+            }
+
+            RoutePatternPart VisitPart(RoutePatternPart part)
+            {
+                if (!part.IsParameter)
+                {
+                    return part;
+                }
+
+                var parameter = (RoutePatternParameterPart)part;
+                var @default = parameter.Default;
+
+                if (updatedDefaults.TryGetValue(parameter.Name, out var newDefault))
+                {
+                    if (parameter.Default != null)
+                    {
+                        var message = Resources.FormatTemplateRoute_CannotHaveDefaultValueSpecifiedInlineAndExplicitly(parameter.Name);
+                        throw new InvalidOperationException(message);
+                    }
+
+                    if (parameter.IsOptional)
+                    {
+                        var message = Resources.TemplateRoute_OptionalCannotHaveDefaultValue;
+                        throw new InvalidOperationException(message);
+                    }
+
+                    @default = newDefault;
+                }
+                
+                if (parameter.Default != null)
+                {
+                    updatedDefaults.Add(parameter.Name, parameter.Default);
+                }
+
+                if (!updatedConstraints.TryGetValue(parameter.Name, out var parameterConstraints) &&
+                    parameter.Constraints.Count > 0)
+                {
+                    parameterConstraints = new List<RoutePatternConstraintReference>();
+                    updatedConstraints.Add(parameter.Name, parameterConstraints);
+                }
+
+                if (parameter.Constraints.Count > 0)
+                {
+                    parameterConstraints.AddRange(parameter.Constraints);
+                }
+
+                return ParameterPartCore(
+                    parameter.Name,
+                    @default,
+                    parameter.ParameterKind,
+                    (IEnumerable<RoutePatternConstraintReference>)parameterConstraints ?? Array.Empty<RoutePatternConstraintReference>());
+            }
+        }
+
+        public static RoutePatternPathSegment Segment(IEnumerable<RoutePatternPart> parts)
+        {
+            if (parts == null)
+            {
+                throw new ArgumentNullException(nameof(parts));
+            }
+
+            return SegmentCore(parts);
+        }
+
+        public static RoutePatternPathSegment Segment(params RoutePatternPart[] parts)
+        {
+            if (parts == null)
+            {
+                throw new ArgumentNullException(nameof(parts));
+            }
+
+            return SegmentCore(parts);
+        }
+
+        private static RoutePatternPathSegment SegmentCore(IEnumerable<RoutePatternPart> parts)
+        {
+            return new RoutePatternPathSegment(parts.ToArray());
```

Comment:
`parts` could sometimes already be an array (I don't know how deep you want to go on perf at the moment)

Reply:
I'd still want to make a defensive copy if the user called `Segment(myArray)`